platform-sdk/docs/proposals/binary-state-api/proposal.md

```diff
@@ -0,0 +1,351 @@
+# Design Proposal: Introducing a Low-Level State API and Refactoring the Existing State API
+
+---
+
+## Summary
+
+This proposal outlines the introduction of a new low-level State API and the refactoring of the current State API. The primary goal is to establish a clean separation between the State API and application logic, facilitating seamless integration with the Block Node.
+
+## Purpose and Context
+
+The current State API has several critical design flaws:
+
+* It depends on the HAPI module, which impedes integration with the Block Node due to differences in protobuf protocol definitions.
+* It includes interfaces and logic that are tightly coupled to the application layer but do not belong in the state management layer.
+
+Previously, before the [typeless Virtual Map](https://github.com/hiero-ledger/hiero-consensus-node/issues/16785) implementation, we had to rely on service schemas to obtain codecs for key and value types. With that limitation removed, we now have the opportunity to introduce a clearer separation of concerns and eliminate unnecessary dependencies—such as the dependency on HAPI.
+
+## Architecture
+
+### Current class diagram
+
+![Current](state-api-current.svg)
+
+### Proposed class diagram
+
+![Proposed](state-api-proposed.svg)
+
+### Java classes
+
+New `State` interface will be defined as follows:
+
+```java
+/**
+ * Represents low-level State API to store singletons, key/value pairs, and queues by their ids.
+ */
+public interface BinaryState {
+    /**
+     * Puts serialized singleton value.
+     *
+     * @param id an id of the singleton type
+     */
+    void putSingleton(int id, Bytes value);
+
+    /**
+     * Puts a singleton value
+     *
+     * @param id    an id of the singleton type
+     * @param codec a codec to convert the object into bytes
+     * @param value value to put
+     */
+    <T> void putSingleton(int id, Codec<T> codec, T value);
+
+    /**
+     * Gets serialized singleton value.
+     *
+     * @param id an id of the singleton type
+     * @return The value, or null if there is no value.
+     */
+    Bytes getSingleton(int id);
+
+    /**
+     * Gets serialized singleton value.
+     *
+     * @param id an id of the singleton type
+     * @param codec a codec to convert the bytes into an object
+     * @return The value, or null if there is no value.
+     */
+    <T> T getSingleton(int id, Codec<T> codec);
```

Comment:
If we decide to leave these methods with codecs (see my previous comment), they should throw `ParseException`, as this is what `Codec.parse(Bytes)` does